opensrp-app/src/test/java/org/smartregister/login/presenter/BaseLoginPresenterTest.java

```diff
@@ -0,0 +1,137 @@
+    @Mock
+    private ViewTreeObserver.OnGlobalLayoutListener layoutListener;
+
+    private AppCompatActivity activity;
+
+    @Before
+    public void setUp() {
+        presenter.setLoginModel(loginModel);
+        presenter.setLoginInteractor(loginInteractor);
+        presenter.setLoginView(new WeakReference<>(loginView));
+        activity = Robolectric.buildActivity(BaseLoginActivityImpl.class).create().get();
+        when(loginView.getActivityContext()).thenReturn(activity);
+    }
+
+    @Test
+    public void testOnDestroyShouldCleanUp() {
+        presenter.onDestroy(false);
+        verify(loginInteractor).onDestroy(false);
+        assertNull(presenter.getLoginView());
+        assertNull(Whitebox.getInternalState(presenter, "mLoginInteractor"));
+        assertNull(Whitebox.getInternalState(presenter, "mLoginModel"));
+    }
+
+    @Test
+    public void testGetLoginViewShouldReturnView() {
+        assertEquals(loginView, presenter.getLoginView());
+    }
+
+    @Test
+    public void testAttemptLoginShouldErrorIfAppIsOutdated() {
+        when(loginView.isAppVersionAllowed()).thenReturn(false);
+        presenter.attemptLogin("john", "doe");
+        verify(loginView).showErrorDialog(activity.getString(R.string.outdated_app));
+        verify(loginView).isAppVersionAllowed();
+        verify(loginView).getActivityContext();
+        verifyNoMoreInteractions(loginView);
+    }
+
+
+    @Test
+    public void testAttemptLoginShouldNotInvokeLoginAndDisplaysErrors() {
+        when(loginView.isAppVersionAllowed()).thenReturn(true);
+        presenter.attemptLogin("", "");
+        verify(loginView).setPasswordError(R.string.error_invalid_password);
+        verify(loginView).setUsernameError(R.string.error_field_required);
+        verify(loginView).enableLoginButton(true);
+        verify(loginInteractor, never()).login(any(), any(), any());
+    }
+
+    @Test
+    public void testAttemptLoginShouldInvokeLogin() {
+        when(loginView.isAppVersionAllowed()).thenReturn(true);
+        presenter.attemptLogin("john", "doe");
+        verify(loginView, never()).setPasswordError(R.string.error_invalid_password);
+        verify(loginView, never()).setUsernameError(R.string.error_field_required);
+        verify(loginView, never()).enableLoginButton(true);
+        verify(loginInteractor).login(any(), eq("john"), eq("doe"));
+    }
+
+    @Test
+    public void testIsUserLoggedOutShouldReturnModelValue() {
+        assertTrue(presenter.isUserLoggedOut());
+        verify(loginModel).isUserLoggedOut();
+    }
+
+    @Test
+    public void testPositionViewsAndCanvasGlobalLayoutListenerProcessor() {
+        presenter.positionViews();
+        final ScrollView canvasSV = loginView.getActivityContext().findViewById(R.id.canvasSV);
+        presenter.canvasGlobalLayoutListenerProcessor(canvasSV, layoutListener);
+
+        RelativeLayout view = loginView.getActivityContext().findViewById(R.id.login_layout);
```

Comment:
There is no way to confirm `global layout listeners ` since `ViewTreeObserver` is final and does not a default constructor.

`android.view.ViewTreeObserver$CopyOnWriteArray` which holds `mOnGlobalLayoutListeners` is also package private and there is not way one can use it or stub it

Reply:
Cool